### ISSUE TITLE
TypeError: that.fieldsColumnsIndex is undefined

### ISSUE THREAD

**Issue by EhsanFahami** [wenzhixin/bootstrap-table#3505]
Hi
I included the bootstrap-table-filter-control.js v2.1.2 and .. .css  and the settings inside the constructor. I've encountered this problem ;
`TypeError: that.fieldsColumnsIndex is undefined`
`bootstrap-table-filter-control.js:196:21`

`columns:[{
                    "field": "name",
                    "title": "name",
                    "sortable": true,
                    "filterControl": 'input',
                    "editable" : {
                        "type" : "text",
                        "title" : "name",
                        validate: function (value) {
                            value = $.trim(value);
                            if (!value) {
                                return 'empty';
                            }
                            var data = $table.bootstrapTable('getData'),
                                index = $(this).parents('tr').data('index');
                            // console.log(data[index]);
                            return '';
                        }
                    }]`

and also set option `filterControl:true` in the constructor.

and also set html 
` <table id="table"
                   data-url="./api/demo.json"
                   data-toggle="table">
            </table>`

I did not know how to code in jsfiddle. So I could not handle your recommendation, and I hope that the same resolution will solve the problem.
what's the solution ?

**Comment by jimgit** [comment]
[I suggest you get a fiddle account ](http://www.dummies.com/web-design-development/javascript/how-to-create-a-jsfiddle-account/)-  it's free.  Then find a working example that is close to your problem.  Then click the FORK button to make it yours.

The reason we ask for fiddle examples is that it saves much time for people that are trying to help you and to isolate the problem to exactly your problem.  Otherwise, we spend too much time setting up your example,  since everyone is volunteering - you need to do your part and put your problem into jsfiddle.

**Comment by djhvscf** [comment]
Are you using the latest version of bootstrap-table?

**Comment by EhsanFahami** [comment]
@jimgit  Thank you for your help

I use the latest version of the bootstrap-table.js  1.11.1
And so I built a jsfiddle to the following address:  [https://jsfiddle.net/EhsanFahami/5qsu84tg/7/](url) 

But I could not accurately create a jsfiddle for bootstrap-table-filter-control , and there are error that prevent the plugin from running.
I have put my custom code there and are visible and probably will be effective to guide me.
**Finally, I want to have a filter table in different modes : select , input:text , input dateTime**

> locale json :  **I put the local json code in the end of the javascript code**


I really appreciate the guidance you give me.

**Comment by jimgit** [comment]
when I run your jsfiddle - I do not get any json data - put a small sample of your data in [http://myjson.com/](http://myjson.com/ )  If you are using chrome, hit the F12 key to goto the chrome debugger.  Make sure you get rid of all the errors in the console tab  by fixing your code or data source.

![image](https://user-images.githubusercontent.com/1921941/34110379-200bce62-e3cc-11e7-9a3c-a21aca0f4b95.png)

**Comment by EhsanFahami** [comment]
Now I do the things you said. But the table still does not work, and it's now that no errors have been recorded in the console.
And I had to delay the implementation of the data table.
I really do not know what to do;
new update from jsfiddle : [jsfiddle.net/EhsanFahami/5qsu84tg/15/](url)

**Comment by jimgit** [comment]
I would suggest starting with a basic table from your json url and working up from there to find the source of the problem.

[Here's a fiddle with your data](http://jsfiddle.net/jimkennelly/e3nk137y/17950/)

**Comment by EhsanFahami** [comment]
@jimgit  I am thankful for your guidance and cooperation, but my problem has not yet been resolved.
Now I put all that I put in my code inside the jsfiddle and there is no error inside the console. even though I put the filter file and made the corresponding settings, I do not see any effect.
**ٔNOTE  that I put my exact code inside jsfiddle.**

[Here's exactly my own code inside the jsfiddle](https://jsfiddle.net/EhsanFahami/g29mf8xL/)

**Comment by jimgit** [comment]
Is this what you are trying to do -[ jsFiddle added filter control ?:](https://jsfiddle.net/jimkennelly/1khjukqx/)


![image](https://user-images.githubusercontent.com/1921941/34175501-3b9608ae-e4c2-11e7-8bbe-2bd415c6a410.png)

**Comment by EhsanFahami** [comment]
Yes, that's precisely what I was looking for and thank you very much for your guidance .
But another problem is that the Farsi search is unsuccessful
I am persian .
The Farsi language is written from right to left, and this is exactly the opposite of English.

Do you know the solution to this problem?

**Comment by EhsanFahami** [comment]
I mistakenly clicked on the close button.
And that's because I'm not familiar with the editor

**Comment by jimgit** [comment]
To search "reverse"  - you may want to setup the SERVER side pagination so the search term is sent to the server - then you have complete control of how the search works since you have to write the search algorithm.  For me - I used entity framwork and my server side looks like:

db.MyDataTables.Where(function(t) t.SearchField.contains("The search Text").skip(x).take(y)

I would do a REVERSE

    Function Reverse(ByVal value As String) As String
        ' Convert to char array.
        Dim arr() As Char = value.ToCharArray()
        ' Use Array.Reverse function.
        Array.Reverse(arr)
        ' Construct new string.
        Return New String(arr)
    End Function

db.MyDataTables.Where(function(t) t.SearchField.REVERSES.contains("The search Text").skip(x).take(y)

**Comment by jimgit** [comment]
You might want to post the REVERSE question as a new issue for people to find it in the future.

**Comment by jimgit** [comment]
Also look at the [data-sorter function JSFIDDLE](http://jsfiddle.net/e3nk137y/1434/)

**Comment by jimgit** [comment]
Wondering if you could modify the search function to something like this in the onSearch  function (in bootstrapTable.js)


 this.searchText = text.split("").reverse().join("");
 this.options.searchText = text.split("").reverse().join("");




![image](https://user-images.githubusercontent.com/1921941/34189837-1a9c04e6-e503-11e7-8fde-565429e8cf32.png)

**Comment by djhvscf** [comment]
@EhsanFahami is this issue resolved? I'm confused with the comments

**Comment by EhsanFahami** [comment]
@jimgit I thank you for your guidance. I really appreciate you :+1: :100: 

@djhvscf 
But about the problem :
Yes, this problem has been solved and now I have no error and the table works correctly.
When I added `<th>` to the `<table>`. I was able to apply the settings and solve my problem.
also I've updated all the plugins and used their latest releases .

Thanks for creating the table from its creators. :+1: 
This plugin is really perfect for the client side :1st_place_medal:
